### PR TITLE
Mk/add phi api

### PR DESCRIPTION
**Pull request by mkangia**
[API details](https://drive.google.com/file/d/1rqJ4dd4XbDhTlsnMmfMhJvgIsY7_QYz-/view)

This is step 1 to have the client available to hook for the integration.
Simple interface for the three end points available.

### CODE REVIEW

File: custom/icds/integrations/clients/phi.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            'apikey': self.apikey,
+            'content-type': 'application/json'
+        }
+
+    def send_request(self):
+        return requests.post(
+            URLS[self.request_type],
+            data=self.data,
+            headers=self._headers
+        )
```
Comment (esoergel) [inline custom/icds/integrations/clients/phi.py]: nit: I think this class would be simpler to read and think about as a function:
```python
def send_phi_api_request(apikey, password, request_type, data):
    return requests.post(
        URLS[request_type],
        data=data,
        headers={
            'Authorization': 'Basic {}' % password,
            'apikey': apikey,
            'content-type': 'application/json'
        }
    )
```

File: custom/icds/integrations/interfaces/phi.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+class BasePHIInterface(object):
+    client_type = None
+
+    def __init__(self):
+        self._response = None
+
+    def _ensure_credentials(self):
+        if not self._api_key or not self._password:
+            raise MissingCredentials("Please set api key and password for phi")
```
Comment (esoergel) [inline custom/icds/integrations/interfaces/phi.py]: This method isn't called anywhere.  Why not move the API key and password (and the check) into the client?  It can raise an exception there, then this interface won't need to know about them.

File: custom/icds/integrations/interfaces/phi.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    @property
+    def _data(self):
+        raise NotImplementedError
+
+    def send_request(self):
+        self._response = self._client.send_request()
+        if self._response.status_code == 403:
+            raise AuthenticationFailure("Authentication Failed")
+        return self._parse_response()
```
Comment (esoergel) [inline custom/icds/integrations/interfaces/phi.py]: `_parse_response()` is temporally coupled to `send_request` - it has a dependency on state from `send_request`.  You can make this dependency explicit by passing `_response` as a parameter to `_parse_response` rather than mutating the class in the middle.

**Review by esoergel — COMMENTED** [review]
I think this could be simplified a lot if you dropped the classes and just used functions.  Overriding class methods like this obscures control flow, and makes a pretty straightforward bit of logic hard to follow.  I could imagine, for instance, something as simple as:

```python
def validate_phi(phi_id):
    response = send_phi_api_request('validate', {'phi_id': phi_id})
    return response.get('result') == 'true'

# TODO similar functions for search and get

def send_phi_api_request(request_type, data):
    if not settings.PHI_API_KEY and settings.PHI_PASSWORD:
        raise 
    return requests.post(
        URLS[request_type],
        data=data,
        headers={
            'Authorization': 'Basic {}' % settings.PHI_PASSWORD,
            'apikey': settings.PHI_API_KEY,
            'content-type': 'application/json'
        }
    ).json()
```

File: custom/icds/integrations/interfaces/phi.py (hunk 2 above)
Comment (mkangia) [reply custom/icds/integrations/interfaces/phi.py]: ah, missed to call that in `__init__`.

Ya i think having it in the interface just makes it fail earlier rather than waiting for the call through the client to happen.

File: custom/icds/integrations/interfaces/phi.py (hunk 3 above)
Comment (mkangia) [reply custom/icds/integrations/interfaces/phi.py]: i see. i did that earlier.
I did store the raw response on `_response` so that it can be referred later if needed. and then it made sense to just use it instead of passing it to the method. So you are suggesting, i store it on `_response` but still pass it as an argument?

**Comment by mkangia** [comment]
@esoergel I personally am in favour of classes here. Agree that methods would be less code but it would then be a bunch of methods with no clear connection to each other whereas class would keep those related methods together categorising them well on the basis of type of request.

File: custom/icds/integrations/interfaces/phi.py (hunk 2 above)
Comment (mkangia) [reply custom/icds/integrations/interfaces/phi.py]: https://github.com/dimagi/commcare-hq/pull/25480/commits/44ca7cc2103fb37f8b6e40540d619cf5b21109cb

File: custom/icds/integrations/interfaces/phi.py (hunk 3 above)
Comment (esoergel) [reply custom/icds/integrations/interfaces/phi.py]: I think it'd be better to split it up.  One problem is setting up and making the request, and one is dealing with the result.  That request makes a logical breaking point.  If one class deals with both things, then half of the class only applies to before the request, while the other half only applies to after the request.  You can split it up to one unit of code (IMO a function would be better) to prepare the request, and one to interact with the result.

Think of how `requests` works - you pass the parameters to a function, which does the action and returns a separate "response" object.  That object can be interacted with in multiple ways if necessary.  Right now this is probably simple enough to just return the parsed JSON, but if the needs became more complicated, I could see wrapping that in a custom class which knows how to access various parts of the response.

File: custom/icds/integrations/interfaces/phi.py (hunk 2 above)
Comment (esoergel) [reply custom/icds/integrations/interfaces/phi.py]: I'm not convinced that distinction matters much - you can still do the check before any network requests.  The reason I'm belaboring this point is that I'd expect the division to be:
"client" code - interacts with the external endpoint, handles network stuff
"interface" code - presents a simple, minimal interface for other code which should only know about this API in an abstract sense.

**Comment by esoergel** [comment]
> @esoergel I personally am in favour of classes here. Agree that methods would be less code but it would then be a bunch of methods with no clear connection to each other whereas class would keep those related methods together categorising them well on the basis of type of request.

What related methods?  Each of these classes has only one public method.  If it's a question of variable scoping, you can use the file itself as the namespace.  You could make these all staticmethods (or methods if necessary) of a class too if you wanted - say `PHIInterfaces.validate_phi`, for instance.

My issue is with using inheritance and inverted control flow when simple function calls would do the same thing much more cleanly.  You can't inspect `ValidatePHI` alone and know how to interact with it.  You have to jump back and forth between that and `BasePHIInterface` for that.

**Comment by mkangia** [comment]
thanks for the feedback @esoergel 
Removed class and brought it down just to a method.

I guess i would let the calling functions deal with how the response is to be parsed like you said so removed that code.

**Comment by esoergel** [comment]
Glad I was able to convince you!  My last comment is that I do think it would be useful to have a light wrapper function for each usage of the API which knows about the request and response formats, so the calling code doesn't need to.  It would make it easier to update if the API ever changed, and would be an easier thing to mock in tests.

**Comment by mkangia** [comment]
Thanks @esoergel 
I think i am fine with leaving this for now and see how i use them in repeaters. I have started a PR for that, https://github.com/dimagi/commcare-hq/pull/25530.
Made this ready for review.

**Review by esoergel — APPROVED** [review]
> I think i am fine with leaving this for now and see how i use them in repeaters.

Makes a lot of sense.  I'm on board with that.

File: custom/icds/integrations/phi.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+from django.conf import settings
+
+import requests
+
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/integrations/phi.py]: Local isort says okay, even with current cfg in HQ repo so ignoring this.

**Review by kaapstorm — APPROVED** [review]
